src/snowflake/connector/network.py

```diff
@@ -246,6 +248,46 @@ def __call__(self, r):
         return r
 
 
+class SessionPool(object):
+    def __init__(self, rest: "SnowflakeRestful"):
+        # A stack of the idle sessions
+        self._idle_sessions: List[Session] = []
+        self._active_sessions: Set[Session] = set()
+        self._rest: "SnowflakeRestful" = rest
+
+    def get_session(self) -> Session:
+        """Returns a session from the session pool or creates a new one."""
+        try:
+            session = self._idle_sessions.pop()
+        except IndexError:
+            session = self._rest.make_requests_session()
+        self._active_sessions.add(session)
+        return session
+
+    def return_session(self, session: Session) -> None:
+        """Places an active session back into the idle session stack."""
+        try:
+            self._active_sessions.remove(session)
+        except KeyError:
+            logger.debug("session doesn't exist in the active session pool. Ignored...")
+        self._idle_sessions.append(session)
+
+    def __repr__(self):
+        return f"Active request sessions: {len(self._active_sessions)}, Idle: {len(self._idle_sessions)}"
```

Comment:
I like to put the class name in the beginning of these, the rest of the change is just a different idea (you don't have to actually accept this).
```suggestion
        return f"SessionPool {len(self._active_sessions)}/{{len(self._active_sessions)} + len(self._idle_sessions)} active sessions"
```